sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.plans.logical.statsEstimation.FilterEstimation
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * A rule that reorder predicate expressions to get better performance.
+ */
+object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
+
+  // Return 1.0D - 1.0D / Int.MaxValue if the expression is not supported calculate
+  // filter selectivity to avoid rankingAnd always return 0.0D.
+  private def selectivity(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    filterEstimation.calculateFilterSelectivity(exp, false)
+      .getOrElse(1.0D - 1.0D / Int.MaxValue)
+  }
+
+  // Formula: (selectivity - 1.0D) / expression cost
+  private def rankingAnd(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    (selectivity(exp, filterEstimation) - 1.0D) / expressionCost(exp)
+  }
+
+  // Formula: (-selectivity) / expression cost
+  private def rankingOr(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    -selectivity(exp, filterEstimation) / expressionCost(exp)
+  }
+
+  // The cost of a call expression e is computed as:
+  //   cost(exp) = typeSize + functionCost + cost(children).
```

Comment:
how do we define `functionCost`?

Reply:
The `functionCost` refers to Hive and Impala:
https://github.com/apache/hive/blob/91a92e323a29a49b5ad3b860b9523f347b6b8d4c/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java#L243-L281
https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L79-L94